### ISSUE TITLE
Lektor Build does not Scale with Increased Content Files

### ISSUE THREAD

**Issue by oscarmlage** [lektor/lektor#180]
I've a blog with about 2000 posts and trying to make a simple build it takes about 85mins: `Finished build in 5013.00 sec`. On the other hand, the `lektor server` petitions are very slow, about 20 secs per request. Have taken a look building the site with the `--profile` option to analyze the _/tmp/lektor-_.prof*, I've attached the result graphs (content.jpg and nocontent.jpg at the end of the issue)

The last try I've done was modifying the `EPHEMERAL_RECORD_CACHE_SIZE to 10000000` (I've saved the file, deactivated the environment + activated it again before to run the build, _jic_) and the results are the following:

```
$ lektor build
Started build
Finished build in 3087.89 sec
Started prune
Finished prune in 0.49 sec
```

I'll add more information (to see if it's something related to templates or static files, the only one static file I load - an empty style.css - takes about 10 secs to load or whatever... once I make more research work.

![content](https://cloud.githubusercontent.com/assets/98542/13968468/1f6c2ff4-f07d-11e5-9333-ed1df05cfa39.jpg)
![nocontent](https://cloud.githubusercontent.com/assets/98542/13968467/1f69156c-f07d-11e5-96f8-a2980319bda9.jpg)

**Comment by oscarmlage** [comment]
I've created [a script](https://gist.github.com/oscarmlage/39dec1022db82b543053) to make contents and test, some results (with `EPHEMERAL_RECORD_CACHE_SIZE to 10000000`):

With my custom templates, (blog index, blog pages, blog details, tags, tag pages):

```
$ ./create_content.py --posts 100 --tags 10 ; lektor clean ; lektor build
Finished build in 8.93 sec
$ ./create_content.py --posts 200 --tags 10 ; lektor clean ; lektor build
Finished build in 25.73 sec
$ ./create_content.py --posts 500 --tags 10 ; lektor clean ; lektor build
Finished build in 142.04 sec
$ ./create_content.py --posts 1000 --tags 10 ; lektor clean ; lektor build
Finished build in 673.77 sec
```

With default _quickstart_ templates (blog index, blog pages, blog details, tags):

```
$ ./create_content.py --posts 100 --tags 10 ; lektor clean ; lektor build
Finished build in 2.60 sec
$ ./create_content.py --posts 200 --tags 10 ; lektor clean ; lektor build
Finished build in 6.33 sec
$ ./create_content.py --posts 500 --tags 10 ; lektor clean ; lektor build
Finished build in 18.13 sec
$ ./create_content.py --posts 1000 --tags 10 ; lektor clean ; lektor build
Finished build in 87.12 sec
```

**Comment by oscarmlage** [comment]
I've created [a script](https://gist.github.com/oscarmlage/39dec1022db82b543053) to make contents and test, some results (with `EPHEMERAL_RECORD_CACHE_SIZE to 10000000`):

With my custom templates, (blog index, blog pages, blog details, tags, tag pages):
```
$ ./create_content.py --posts 100 --tags 10 ; lektor clean ; lektor build
Finished build in 8.93 sec
$ ./create_content.py --posts 200 --tags 10 ; lektor clean ; lektor build
Finished build in 25.73 sec
$ ./create_content.py --posts 500 --tags 10 ; lektor clean ; lektor build
Finished build in 142.04 sec
$ ./create_content.py --posts 1000 --tags 10 ; lektor clean ; lektor build
Finished build in 673.77 sec
```

With default *quickstart* templates (blog index, blog pages, blog details, tags):
```
$ ./create_content.py --posts 100 --tags 10 ; lektor clean ; lektor build
Finished build in 2.60 sec
$ ./create_content.py --posts 200 --tags 10 ; lektor clean ; lektor build
Finished build in 6.33 sec
$ ./create_content.py --posts 500 --tags 10 ; lektor clean ; lektor build
Finished build in 18.13 sec
$ ./create_content.py --posts 1000 --tags 10 ; lektor clean ; lektor build
Finished build in 87.12 sec
```

**Comment by mitsuhiko** [comment]
I already mentioned this on gitter but Lektor in the current state just does not really scale to large counts within a folder. I'm not sure yet which path to go down to improve this but sooner or later the system will have to build indexes in sqlite and cut down on IO calls. But that's more work and the question is if there are some lower hanging fruits we can tackle before that is necessary.

**Comment by mitsuhiko** [comment]
What you can see from the times you took is that this is literally a scaling problem: https://www.wolframalpha.com/input/?i=%5B%5B100,+9%5D,+%5B200,+26%5D,+%5B500,+142%5D,+%5B1000,+674%5D%5D

**Comment by oscarmlage** [comment]
Well, I think times as the ones I got with the default templates are more than reasonable, the key here is to know what is happening between default templates and mines. After making some more try and failure  work, I've realized that the line `{{ render_nav_bottom(this) }}` in _post-detail_ template makes the build process really slow, this macro calculates previous/next siblings posts:

```
{% macro render_nav_bottom(record) %}
  <nav class="nav-below">
    <div class="nav-previous">
      {% set siblings = record.get_siblings() %}
      {% set older = siblings.next_page %}
      {% set newer = siblings.prev_page %}
      {% if older %}
        <a href="{{ older|url }}"><span class="meta-nav">&larr;&nbsp;</span>{{ older.title }}</a>
      {% endif %}
    </div>
    <div class="nav-next">
      {% if newer %}
        <a href="{{ newer|url }}">{{ newer.title }}<span class="meta-nav">&nbsp;&rarr;</span></a>
      {% endif %}
    </div>
  </nav>
{% endmacro %}
```

Disallowing that macro times are more similar to the default ones. Something similar happens with the pagination of tags too - not that critical in this case -. I wonder if the key here is way paginations are being calculated. More information about tag pagination, models/tag.ini:

```
[model]
name = Tag
label = {{ this.name }}
hidden = yes

[children]
replaced_with = site.get('/').pagination.items.filter(F.tags.contains(this._id))

[fields.name]
label = Name
type = string
size = large

[pagination]
enabled = yes
per_page = 10
```

macro `{{ render_pagination(this.pagination) }}`:

```
{% macro render_pagination(pagination) %}
  <div class="pagination">
    {% if pagination.has_prev %}
      <a href="{{ pagination.prev|url }}">&laquo; Previous</a>
    {% else %}
      <span class="disabled">&laquo; Previous</span>
    {% endif %}
    | {{ pagination.page }} |
    {% if pagination.has_next %}
      <a href="{{ pagination.next|url }}">Next &raquo;</a>
    {% else %}
      <span class="disabled">Next &raquo;</span>
    {% endif %}
  </div>
{% endmacro %}
```

**Comment by mitsuhiko** [comment]
Generally right now the dependency tracking is on the entirety of the record which is not ideal. It rebuilds way too much.  Ideally we could track on a field level.  That would help siblings a lot.

**Comment by tgpfeiffer** [comment]
Just wondering, because it's been a while, does this problem still exist?

**Comment by jcharaoui** [comment]
Yes. We have several sites that take tens of minutes to build. Examples:

- https://gitlab.torproject.org/tpo/web/blog/-/jobs/388687
- https://gitlab.torproject.org/tpo/web/tpo/-/jobs/389286

**Comment by kno10** [comment]
Is there an easy way to bypass this dependency tracking?

**Comment by johnzhou721** [comment]
Hi! How do you set ``EPHEMERAL_RECORD_CACHE_SIZE``?
